src/async.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright 2024 Adobe
+// All Rights Reserved.
+//
+// NOTICE: Adobe permits you to use, modify, and distribute this file in accordance with the terms
+// of the Adobe license agreement accompanying it.
+
+// identity
+#include "orc/async.hpp"
+
+// stdc++
+#include <cassert>
+
+// application
+#include "orc/settings.hpp"
+#include "orc/task_system.hpp"
+#include "orc/tracy.hpp"
+
+/**************************************************************************************************/
+
+namespace orc {
+
+/**************************************************************************************************/
+
+namespace {
+
+/**************************************************************************************************/
+
+struct work_counter {
+    struct state {
+        void increment() {
+            {
+                std::lock_guard<std::mutex> lock(_m);
+                ++_n;
+            }
+            _c.notify_all();
+        }
+
+        void decrement() {
+            {
+                std::lock_guard<std::mutex> lock(_m);
+                --_n;
+            }
+            _c.notify_all();
+        }
+
+        void wait() {
+            std::unique_lock<std::mutex> lock(_m);
+            if (_n == 0) return;
+            _c.wait(lock, [&] { return _n == 0; });
+        }
+
+        std::mutex _m;
+        std::condition_variable _c;
+        std::size_t _n{0};
+    };
+
+    using shared_state = std::shared_ptr<state>;
```

Comment:
I think removing the `using` and using `std::shared_ptr` directly would be clearer in this very limited scope - this is private, inline, and small.

Reply:
I am inclined to disagree, as there is a single type here whose bespoke name makes the rest of the code more readable. At the same time it's not a hill I'm willing to die on, so will change it. (This code has also been in ORC for some time - the only reason it's in this PR is because I moved it from its original location to this file.)